test/unit/inbound/process-submit-payment-request.test.js

```diff
@@ -0,0 +1,349 @@
+const mockCommit = jest.fn()
+const mockRollback = jest.fn()
+const mockTransaction = {
+  commit: mockCommit,
+  rollback: mockRollback
+}
+
+jest.mock('../../../app/data', () => {
+  return {
+    sequelize:
+       {
+         transaction: jest.fn().mockImplementation(() => {
+           return { ...mockTransaction }
+         })
+       }
+  }
+})
+
+jest.mock('../../../app/inbound/get-payment-request-by-invoice-number')
+const getPaymentRequestByInvoiceNumber = require('../../../app/inbound/get-payment-request-by-invoice-number')
+
+jest.mock('../../../app/inbound/save-invoice-number')
+const saveInvoiceNumber = require('../../../app/inbound/save-invoice-number')
+
+jest.mock('../../../app/inbound/save-payment-request')
+const savePaymentRequest = require('../../../app/inbound/save-payment-request')
+
+jest.mock('../../../app/inbound/save-invoice-lines')
+const saveInvoiceLines = require('../../../app/inbound/save-invoice-lines')
+
+const processSubmitPaymentRequest = require('../../../app/inbound')
+
+let paymentRequest
+
+describe('process submit payment request', () => {
+  beforeEach(() => {
+    paymentRequest = JSON.parse(JSON.stringify(require('../../mockPaymentRequest').submitPaymentRequest))
+
+    getPaymentRequestByInvoiceNumber.mockReset()
+    getPaymentRequestByInvoiceNumber
+      .mockResolvedValueOnce(null)
+      .mockResolvedValueOnce(paymentRequest)
+    getPaymentRequestByInvoiceNumber.mockReturnValue(null)
+    saveInvoiceNumber.mockReturnValue(undefined)
+    savePaymentRequest.mockReturnValue(undefined)
+    saveInvoiceLines.mockReturnValue(undefined)
+  })
+
+  afterEach(() => {
+    jest.clearAllMocks()
+    // jest.resetAllMocks()

```

Comment:
Ssoooo what we could do to resolve the issue defined in the novel apart is: `jest.resetAllMocks()`

This will reset that 'loop pointer' back to 1st time: `null`, 2nd time: `paymentRequest`.

However, doing so means that _all_ of the mocks not defined within `beforeEach` would be `undefined` (because they were `reset`) for the 2nd test function.

Question is: what is the best development approach?

Reply:
I don't think `resetAllMocks` is the right approach, otherwise you'd have to move all the mock definitions into the `beforeEach/afterEach` and also declare them as `let` outside so they can be accessed in each test scope.  All feels a bit more messy.

I'd probably reset that particular mock in the `beforeEach/afterEach` so you don't need to keep repeating it if the number of tests increase as you have